### PR TITLE
Separate optimization methods from CDIModel into separate class Reconstructors

### PR DESCRIPTION
**Pull request by yoshikisd**
# PR Summary
This PR separates ptychography optimization/reconstruction methods from `CDIModel` to a new class called `Reconstructor`. 

This separation is intended to separate optimization-related code from `CDIModel` into a dedicated class `Reconstructor`. The motivation of this change was to make the CDTools more compatible with other PyTorch-related packages which expect libraries/scripts to follow specific code patterns. Specifically, problems arise if optimization is a method of `torch.nn.module`/`cdtools.models.CDIModel`, which was the case as of `v0.3.0.pypi1`. Additionally, `Reconstructor`'s implementation allows us to re-use an optimizer rather than create a brand new one each time a new reconstruction loop is called (as was done with calls to `model.{Adam, LBFGS, SGD}_optimize`)

This PR serves to split PR https://github.com/cdtools-developers/cdtools/pull/17, a development PR that ended up creating both `Reconstructors` as well as multi-GPU capability. This PR was created by checking out all `Reconstructors`-related stuff from PR https://github.com/cdtools-developers/cdtools/pull/17 and omits any multi-GPU related code. 

@allevitan If you can add your reviews from https://github.com/cdtools-developers/cdtools/pull/17, that would be appreciated.

# Technical details

## Implementation:

`cdtools.reconstructors` is largely based on methods from `cdtools.models.CDIModel`. For the base class `cdtools.reconstructors.Reconstructor`, the method(s):
- `_run_epoch` and `optimize` are based on `cdtools.models.CDIModel.AD_Optimize`.
- `setup_dataloader` is based on similar blocks of data-loading code present in `cdtools.models.CDIModel.{Adam, LBFGS, SGD}_optimize`
- `adjust_optimizer` is not implemented and must be defined within the subclasses to adjust the optimizer hyperparameters.

For the subclasses `cdtools.reconstructors.{Adam, LBFGS, SGD}`, the method(s):
- `__init__` initializes an `optimizer` attribute as a `torch.optim.{Adam, LBFGS, SGD}` using the model parameters.
-  `adjust_optimizer` changes the relevant optimizer hyperparameters
- `optimize` are based on `cdtools.models.CDIModel.{Adam, LBFGS, SGD}

The main change to the code behavior is that the hyperparameters are updated between different reconstruction loops without creating a new `torch.optim.{Adam, LBFGS, SGD}`. In some cases, this can result in an improvement in reconstruction quality as discussed in (https://github.com/cdtools-developers/cdtools/pull/17#issuecomment-2896536130).

## Changes in `cdtools.models.CDIModel` and backwards compatibility

The following changes have been made in `cdtools.models.CDIModel` methods:
- `AD_Optimize` has been removed.
- `{Adam, LBFGS, SGD}_Optimize` now initialize a `reconstructor` attribute as a `cdtools.reconstructors.{Adam, LBFGS, SGD}` and make a call to their `optimize` method.

The latter change allows scripts written using `v0.3.0.pypi1` to be backwards compatible with this update. 

However, this comes at the cost of creating a circular dependency between `cdtools.models.CDIModel` and `cdtools.reconstructors`. From a reconstruction standpoint it doesn’t seem to cause any issues. However, it’s not clear if this may cause issues in the long term in terms of maintainability. Thus, we should consider deprecating `{Adam, LBFGS, SGD}_Optimize` in favor of the following code pattern:
```diff
  import cdtools

  filename = ...
  dataset = ...
  model = cdtools.models.FancyPtycho.from_dataset(...)

  device = 'cuda'
  model.to(device=device)
  dataset.get_as(device=device)

+ recon = cdtools.optimizer.Adam(model, dataset)

- for loss in model.Adam_optimize(50, dataset, lr=0.02, batch_size=10):
+ for loss in recon.optimize(50, lr=0.02, batch_size=10):
    print(model.report())
    if model.epoch % 10 == 0:
        model.inspect(dataset)
 ```

## PyTest

Three tests have been created in test_reconstructors.py which function similar to the tests for the models. These three tests run reconstruction loops using both the traditional method (model.XXX_Optimize) and new method (recon.optimize), but use different optimizers (Adam, LBFGS, SGD), datasets (gold ball dataset for Adam and SGD, siemens star optical data for LBFGS), and models (FancyPtycho for Adam and SGD, RPI for LBFGS). These three tests check the following:
- Calls to `Reconstructor.adjust_optimizer` are updating the hyperparameters
- Ensure that the model attribute of the `Reconstructor` points to the original model it was provided as an initialization parameter
- Ensure that reconstructions performed by the old and new methods produce exactly identical results given the same random seed
- Ensure that the final loss of the reconstructions remain below a specified threshold

Additionally, the first test `test_Adam_gold_balls` checks if data loading methods specific to single-GPU operations are working as intended.

### CODE REVIEW

File: src/cdtools/reconstructors/base.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        initializing the Reconstructor subclass.
+    - **scheduler** -- A `torch.optim.lr_scheduler` that may be defined during
+        the `optimize` method.
+    - **data_loader** -- A torch.utils.data.DataLoader that is defined by
+        calling the `setup_dataloader` method.
+    """
+    def __init__(self,
+                 model: CDIModel,
+                 dataset: CDataset,
+                 subset: Union[int, List[int]] = None):
```
Comment (yoshikisd) [inline src/cdtools/reconstructors/base.py:53]: In PR https://github.com/cdtools-developers/cdtools/pull/17, it was suggested that subset could be moved into the optimize function call. More generally, the though we had was that a subset call shouldn't only be made within __init__. 

One question I have @allevitan is do you foresee a case where subset will be changed several times within a given reconstruction script? I think that doing so would require us to have dataset as a parameter for `Reconstructor.optimize` calls, which is fine IMO. 

On a similar note, it makes the most sense to me to have dataset/dataloader-related definitions contained within `Reconstructor.setup_dataloader`.

File: src/cdtools/models/base.py [hunk 2]
```diff
@@ -521,177 +345,206 @@ def Adam_optimize(
...
+        #
+        # The only way to do this is to make the Adam reconstructor an
+        # attribute of the model. But since the Adam reconstructor also
+        # depends on CDIModel, this seems to give rise to a circular import
+        # error unless we import cdtools.reconstructors within this method:
+        if not hasattr(self, 'reconstructor'):
+            from cdtools.reconstructors import AdamReconstructor
+            self.reconstructor = AdamReconstructor(model=self,
+                                                   dataset=dataset,
+                                                   subset=subset)
```
Comment (yoshikisd) [inline src/cdtools/models/base.py:390]: In PR https://github.com/cdtools-developers/cdtools/pull/17#discussion_r2245706465, the following comment was made:

> This needs to be updated so that if the user changes which optimizer they're using, it doesn't pass a silent error. I would lean towards just creating a new optimizer for each new function (less good behavior, but preserving old behavior for old scripts), but I could be swayed for sure. What do other folks think?

I think this change is fine and can be made with, for example

```python
from cdtools.reconstructors import AdamReconstructor
if not hasattr(self, 'reconstructor') or \
        not isinstance(self.reconstructor, AdamReconstructor):
    self.reconstructor = AdamReconstructor(model=self,
                                           dataset=dataset,
                                           subset=subset)
```

This point also brings up a different thought: what should the code pattern look like if we make switches between optimizers convenient if we 100% switch to using `Reconstructor.optimizer` from `CDIModel.XXX_optimize`. Currently, the following pattern would be required when using `Reconstructor':

```python
# Here we use Adam first
recon = cdtools.reconstructors.AdamReconstructor(model, dataset)
for loss in recon.optimize(...):
    ...

# If we want to switch to something else, we need to re-define the reconstructor
# or create a new one (e.g., called recon2)
recon = cdtools.reconstructors.LBFGSReconstructor(model, dataset)
for loss in recon.optimize(...):
    ...
```
This pattern is not too bad IMO and gives the user the ability to switch between previously-created reconstructors.

**Review by yoshikisd — COMMENTED** [review]
I've addressed some of the reviews requested in PR https://github.com/cdtools-developers/cdtools/pull/17.

Comments https://github.com/cdtools-developers/cdtools/pull/17#discussion_r2245757695 and https://github.com/cdtools-developers/cdtools/pull/17#discussion_r2245636032 were addressed in https://github.com/cdtools-developers/cdtools/pull/47/commits/82ce845385170afbad9845e73be1d487e17b7373.

**Review by allevitan — APPROVED** [review]
I've rechecked the tests, everything is passing. Ready to merge.
